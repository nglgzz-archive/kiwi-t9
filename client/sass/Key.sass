@import variables


.key
  +flex-center
  background: linear-gradient(to bottom, $athens-gray 40%, $mischka 100%)
  border: 0
  flex-direction: row
  height: 50px
  margin: 5px 10px
  outline: none
  width: calc(100% / 3 - 20px)

  &:active
    background: linear-gradient(to bottom, darken($athens-gray, 10) 40%, darken($mischka, 10) 100%)

  &--left
    +shadow(0, 3, 0, 3)
    border-radius: 60% 90%

  &--right
    +shadow(0, 3, 0, 3)
    border-radius: 90% 60%
    flex-direction: row-reverse

  &--center
    +shadow(3, 3, 3, 3)
    border-radius: 50%
    transform: translateY(40%)


.key-label
  font-size: 1.5em
  font-weight: bold
  margin: 0 5px


+respond-after($mobile)
  .key
    &:hover
      background: linear-gradient(to bottom, lighten($athens-gray, 10) 40%, lighten($mischka, 10) 100%)

    &--left,
    &--right
      &:hover
        +shadow(0, 2, 0, 2)

      &:active
        +shadow(0, 4, 0, 4)

    &--center
      &:hover
        +shadow(2, 2, 2, 2)

      &:active
        +shadow(4, 4, 4, 4)
