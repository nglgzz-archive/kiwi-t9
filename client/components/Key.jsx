import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as Suggestions from 'actions/Suggestions';
import * as Symbols from 'actions/Symbols';
import 'sass/Key.sass';


@connect(store => ({
  digits: store.text.digits,
}))
export default class Key extends Component {
  constructor(props) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    switch (this.props.label) {
      case '1':
        this.props.dispatch(Symbols.insert());
        break;

      case '*':
        // Cycle through suggestions
        this.props.dispatch(Suggestions.next());
        break;

      case '0':
        // Finish the current word and start a new one
        this.props.dispatch(Suggestions.end());
        break;

      case '#':
        // add new word
        break;

      default:
        this.props.dispatch(Suggestions.fetch((
          this.props.digits + this.props.label
        )));
    }
  }

  render() {
    const { label, description, position } = this.props;
    const positionName = ['left', 'center', 'right'][position % 3];

    return (
      <button
        className={`key key--${positionName}`}
        onClick={this.handleClick}
      >
        <span className="key-label">{label}</span>
        <span className="key-description">{description}</span>
      </button>
    );
  }
}


Key.propTypes = {
  label: PropTypes.string.isRequired,
  position: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
};
