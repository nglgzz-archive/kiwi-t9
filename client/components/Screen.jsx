import React, { Component } from 'react';
import { connect } from 'react-redux';
import 'sass/Screen.sass';


@connect(store => ({
  text: store.text.text,
  lastWord: store.text.lastWord,
}))
export default class Screen extends Component {
  render() {
    const { text, lastWord } = this.props;

    return (
      <div className="screen">
        <div className="statusbar">
          {(new Date()).toLocaleTimeString().slice(0, 5)}
        </div>

        <div className="textbox">
          <span>{text}</span>
          <span className="last-word">{lastWord}</span>
          <span className="cursor" />
        </div>

        <div className="actions">
          <a href="/">Send</a>
        </div>
      </div>
    );
  }
}
