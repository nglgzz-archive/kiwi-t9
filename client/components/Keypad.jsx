import React from 'react';
import Key from 'components/Key';
import keypadLayout from 'utils/keypadLayout.json';
import 'sass/Keypad.sass';


export default () => {
  const keys = keypadLayout.map(({ label, description }, index) => (
    <Key
      position={index}
      key={label}
      label={label}
      description={description}
    />
  ));

  return (
    <div className="keypad">{keys}</div>
  );
};
