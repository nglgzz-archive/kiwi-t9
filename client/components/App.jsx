import React from 'react';
import Keypad from 'components/Keypad';
import Screen from 'components/Screen';
import 'sass/App.sass';


export default () => (
  <div className="phone">
    <div className="speaker">
      <span />
      <span />
      <span />
    </div>

    <Screen />
    <Keypad />
  </div>
);
