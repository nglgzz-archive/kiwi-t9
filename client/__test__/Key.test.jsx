import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Key } from 'components/Key';


Enzyme.configure({ adapter: new Adapter() });


describe('Key component', () => {
  const setup = (propOverrides) => {
    const props = {
      label: '1',
      ...propOverrides,
    };

    return shallow(<Key {...props} />);
  };

  it('renders without crashing', () => {
    const wrapper = setup();
    expect(wrapper.length).toBe(1);
  });

  it('dispatches SYMBOL_INSERT and SYMBOL_NEXT', () => {
    const mockDispatch = jest.fn();
    const wrapper = setup({ dispatch: mockDispatch });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(1);
    expect(mockDispatch).lastCalledWith({
      type: 'SYMBOL_INSERT',
    });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(2);
    expect(mockDispatch).lastCalledWith({
      type: 'SYMBOL_NEXT',
    });

    // Fast forward time by one second, so the next time the symbol button is
    // pressed it will insert a new symbol instead of changing the existing one.
    const timeInASec = new Date(Number(new Date()) + 1000);
    const RealDate = Date;
    // eslint-disable-next-line no-global-assign
    Date = function constructor() { return timeInASec; };

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(3);
    expect(mockDispatch).lastCalledWith({
      type: 'SYMBOL_INSERT',
    });


    // Let's not mess with time and set Date back to its original value.
    // https://imgur.com/gallery/nhFFBm3
    // eslint-disable-next-line no-global-assign
    Date = RealDate;
  });

  it('dispatches WORD_NEXT', () => {
    const mockDispatch = jest.fn();
    const wrapper = setup({ label: '*', dispatch: mockDispatch });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(1);
    expect(mockDispatch).toBeCalledWith({
      type: 'WORD_NEXT',
    });
  });

  it('dispatches CHANGE_CASE', () => {
    const mockDispatch = jest.fn();
    const wrapper = setup({ label: '#', dispatch: mockDispatch });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(1);
    expect(mockDispatch).toBeCalledWith({
      type: 'CHANGE_CASE',
    });
  });

  it('dispatches WORD_END', () => {
    const mockDispatch = jest.fn();
    const wrapper = setup({ label: '0', dispatch: mockDispatch });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(1);
    expect(mockDispatch).toBeCalledWith({
      type: 'WORD_END',
      space: true,
    });
  });

  it('dispatches SUGGESTIONS_FETCH', () => {
    const mockDispatch = jest.fn();
    const wrapper = setup({
      label: '3',
      digits: '12',
      dispatch: mockDispatch,
    });

    wrapper.find('button').simulate('click');
    expect(mockDispatch.mock.calls.length).toBe(1);
    expect(mockDispatch).toBeCalledWith({
      type: 'SUGGESTIONS_FETCH',
      payload: '/api/suggestions?q=123',
    });
  });
});
