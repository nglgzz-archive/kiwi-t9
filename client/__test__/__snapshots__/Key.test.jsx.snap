// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key component matches snapshot 1`] = `
<button
  className="key key--center"
  onClick={[Function]}
>
  <span
    className="key-label"
  >
    1
  </span>
  <span
    className="key-description"
  />
</button>
`;
